pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -536,6 +514,31 @@ func (rc *reconciler) reconstructVolume(volume podVolume) (*reconstructedVolume,
 					newMapperErr)
 			}
 		}
+	} else {
```

Comment:
Is having nil mounter safe to return in `reconstructedVolume`?

Reply:
That's goal of this PR. CSI's `NewMounter` would create directory and a json file for a **block** volume, which we don't want. `reconstructedVolume` should have either `mounter` **or** `mapper`, same as regular volumes in DSW.

Reply:
just to make sure, for block volume, mounter is not needed at all.

Reply:
@jingxu97

Thank you for your comment.

> just to make sure, for block volume, mounter is not needed at all.

As I checked operation_executor/operation_generator code paths for block volume, mounter is not directly used in these paths, except [`GenerateExpandInUseVolumeFunc`](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L1633) .
And I guess that this path won’t be reached for block volume, because `FSResizeRequired` won’t be set due to [this check](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L374).

Is my understanding right? @gnufied @mlmhl

Also, could you double check if we should check any other paths deeper? @jsafrane

```
AttachVolume  --> GenerateAttachVolumeFunc --> <Attacher>
DetachVolume --> GenerateDetachVolumeFunc --> <Detacher>
MountVolume --> GenerateMapVolumeFunc --> <BlockVolumeMapper>
UnmountVolume --> GenerateUnmapVolumeFunc --> <BlockVolumeUnmapper>
UnmountDevice --> GenerateUnmapDeviceFunc --> <BlockVolumeUnmapper>

ExpandInUseVolume --> GenerateExpandInUseVolumeFunc --> <Attacher, Mounter>

VerifyVolumesAreAttached --> GenerateBulkVolumeVerifyFunc --> <Attacher>
VerifyVolumesAreAttachedPerNode --> GenerateVolumesAreAttachedFunc --> <Attacher>
VerifyControllerAttachedVolume --> GenerateVerifyControllerAttachedVolumeFunc --> <None>

ReconstructVolumeOperation --> <None>
CheckVolumeExistenceOperation --> <None>
```

Reply:
@mkimuram, thanks for your analysis! It confirms my finding that mounter is not really needed for block volumes.

Reply:
I was checking here, and thought there might be issue. But since it only loops over container.VolumeMounts which does not contain blockvolumes, it should be fine then.
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L146